client/consensus/epochs/src/lib.rs

```diff
@@ -344,36 +344,31 @@ where
 		&self,
 		slot: E::Slot,
 	) -> Option<(Hash, Number, EpochHeader<E>, EpochIdentifierPosition)> {
+		if let Some((h, n, epoch_n)) = &self.next {
+			if epoch_n.start_slot() <= slot {
+				return Some((*h, *n, epoch_n.into(), EpochIdentifierPosition::Regular))
+			}
+		}
+
 		match &self.current {
-			(_, _, PersistedEpoch::Genesis(epoch_0, _))
-				if slot >= epoch_0.start_slot() && slot < epoch_0.end_slot() =>
-				return Some((
-					self.current.0,
-					self.current.1,
-					epoch_0.into(),
-					EpochIdentifierPosition::Genesis0,
-				)),
-			(_, _, PersistedEpoch::Genesis(_, epoch_1))
-				if slot >= epoch_1.start_slot() && slot < epoch_1.end_slot() =>
-				return Some((
-					self.current.0,
-					self.current.1,
-					epoch_1.into(),
-					EpochIdentifierPosition::Genesis1,
-				)),
-			(_, _, PersistedEpoch::Regular(epoch_n))
-				if slot >= epoch_n.start_slot() && slot < epoch_n.end_slot() =>
-				return Some((
-					self.current.0,
-					self.current.1,
-					epoch_n.into(),
-					EpochIdentifierPosition::Regular,
-				)),
-			_ => {},
-		};
-		match &self.next {
-			Some((h, n, epoch_n)) if slot >= epoch_n.start_slot() && slot < epoch_n.end_slot() =>
-				Some((*h, *n, epoch_n.into(), EpochIdentifierPosition::Regular)),
+			(_, _, PersistedEpoch::Regular(epoch_n)) if epoch_n.start_slot() <= slot => Some((
+				self.current.0,
+				self.current.1,
+				epoch_n.into(),
+				EpochIdentifierPosition::Regular,
+			)),
+			(_, _, PersistedEpoch::Genesis(_, epoch_1)) if epoch_1.start_slot() <= slot => Some((
+				self.current.0,
+				self.current.1,
+				epoch_1.into(),
+				EpochIdentifierPosition::Genesis1,
+			)),
+			(_, _, PersistedEpoch::Genesis(epoch_0, _)) if epoch_0.start_slot() <= slot => Some((
+				self.current.0,
+				self.current.1,
+				epoch_0.into(),
+				EpochIdentifierPosition::Genesis0,
+			)),
```

Comment:
Will `Genesis0` be ever returned? If I read the code correctly the middle match-arm covers also the last one.